WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs

```diff
@@ -1,66 +1,114 @@
 
 namespace WalletWasabi.Backend.Filters
 {
 	public class IdempotentActionFilterAttributeImpl : ActionFilterAttribute
 	{
-		private readonly IMemoryCache _cache;
+		private static TimeSpan CacheTimeout { get; } = TimeSpan.FromMinutes(5);
+		private const string ContextCacheKey = "cached-key";
 
 		public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
 			: base()
 		{
-			_cache = cache;
+			ResponseCache = cache;
 		}
 
+		/// <summary>Guards <see cref="ResponseCache"/>.</summary>
+		/// <remarks>Unfortunately, <see cref="CacheExtensions.GetOrCreate{TItem}(IMemoryCache, object, Func{ICacheEntry, TItem})"/> is not atomic.</remarks>
+		/// <seealso href="https://github.com/dotnet/runtime/issues/36499"/>
+		private object ResponseCacheLock { get; } = new();
+
+		/// <remarks>Guarded by <see cref="ResponseCacheLock"/>.</remarks>
+		private IMemoryCache ResponseCache { get; }
+
+		/// <inheritdoc/>
 		public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
 		{
-			var request = context.HttpContext.Request;
+			HttpRequest request = context.HttpContext.Request;
 
 			if (context.ModelState.IsValid)
 			{
-				if (context.ActionArguments.TryGetValue("request", out var model))
+				if (!context.ActionArguments.TryGetValue("request", out object? model))
 				{
-					var cacheKey = GetCacheEntryKey(request.Path, model);
+					throw new InvalidOperationException("Control actions marked as Idempotent must receive a 'request' argument.");
+				}
+
+				string cacheKey = GetCacheEntryKey(request.Path, model);
+				TaskCompletionSource<IActionResult>? cachedResponseTcs = null;
+				TaskCompletionSource<IActionResult>? responseTcs = null;
 
-					if (_cache.TryGetValue<ObjectResult>(cacheKey, out var cachedResponse))
+				lock (ResponseCacheLock)
+				{
+					if (!ResponseCache.TryGetValue(cacheKey, out cachedResponseTcs))
 					{
-						context.Result = cachedResponse;
-						context.HttpContext.Items.Remove("cached-key");
-						return;
+						responseTcs = new();
+						ResponseCache.Set(cacheKey, responseTcs, DateTimeOffset.UtcNow.Add(CacheTimeout));
 					}
+				}
 
-					context.HttpContext.Items["cached-key"] = cacheKey;
+				if (cachedResponseTcs is not null)
+				{
+					try
+					{
+						context.Result = await cachedResponseTcs!.Task.WithAwaitCancellationAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+						context.HttpContext.Items.Remove(ContextCacheKey);
+						return;
+					}
+					catch (OperationCanceledException)
+					{
+						// Failed to get cached response. Continue as if it were a non-cached request.
+					}
 				}
-				else
+
+				context.HttpContext.Items[ContextCacheKey] = cacheKey;
+
+				ActionExecutedContext executed = await next().ConfigureAwait(false);
+
+				// Request failed for some reason, we don't want to hold any other requests that wait for this one request to finish.
+				if (executed.Exception != null && !executed.ExceptionHandled)
```

Comment:
Could this be?

```suggestion
				if (executed.Exception is { ExceptionHandled: false })
```

Anyway, more important question: what does it mean that it has handled? Handled by what? I mean, are you sure than in case it was handled we should not cancel the `responseTcs`? Do you think it could be possible to write a test for this?

Reply:
> Could this be?

I don't think so because `ExceptionHandled` is not a property of `executed.Exception`.

> Anyway, more important question: what does it mean that it has handled? Handled by what? I mean, are you sure than in case it was handled we should not cancel the `responseTcs`? Do you think it could be possible to write a test for this?

So as far as I could find out, `executed.ExceptionHandled` is `false` unless another C# attribute (by attribute I mean something like `[Idempotent]`) handles the exception. I have tested it manually by throwing an exception in code. 

Whether we can write a test, so, I think we can. It may be like 4/10 hard (mostly because I'm not that proficient in ASP.NET yet).